File: 213.打家劫舍-ii.cpp
/*
 * @lc app=leetcode.cn id=213 lang=cpp
 *
 * [213] 打家劫舍 II
 */

// @lc code=start

#include <vector>
#include <algorithm>
class Solution {
public:
    int rob(std::vector<int>& nums) {
        if(nums.empty()) return 0;
        if(nums.size() == 1) return nums[0];
        
        auto f1 = std::make_pair(0, nums[0]);
        auto f2 = std::make_pair(0, nums[1]);
        for(int i = 1; i < nums.size() - 1; ++i){
            auto ff = std::make_pair(std::max(f1.first, f1.second), nums[i] + f1.first);
            f1 = std::move(ff);

            ff.first = std::max(f2.first, f2.second);
            ff.second = nums[i+1] + f2.first;
            f2 = std::move(ff);
        }
        return std::max(std::max(f1.first, f1.second), std::max(f2.first, f2.second));
    }
};
// @lc code=end

File: 213.打家劫舍-ii.rs
/*
 * @lc app=leetcode.cn id=213 lang=rust
 *
 * [213] 打家劫舍 II
 */

// @lc code=start
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        if nums.len() == 1 {
            return nums[0].max(0);
        }
        // 不偷第一间
        let mut f = vec![[0;2];nums.len()];
        // 不偷最后一间
        let mut g = vec![[0;2];nums.len()];
        g[0][1] = nums[0];
    
        for idx in 1..nums.len() {
            f[idx][0] = f[idx - 1][1].max(f[idx - 1][0]);
            f[idx][1] = f[idx - 1][0] + nums[idx];
        }
        for idx in 1..nums.len() - 1 {
            g[idx][0] = g[idx - 1][1].max(g[idx - 1][0]);
            g[idx][1] = g[idx - 1][0] + nums[idx];
        }
    
        f[nums.len() - 1][0].max(f[nums.len() - 1][1]).max(g[nums.len() - 2][0]).max(g[nums.len() - 2][1])
    }
}
// @lc code=end

